File: lumina/packages/core/src/detection/gaze-tracker.ts
/**
 * GazeTracker - Detects vertical gaze direction to prevent false blink detection
 *
 * When a user looks down and back up quickly, the EAR pattern mimics a blink.
 * This tracker detects such gaze shifts by monitoring iris position relative
 * to the eye opening.
 *
 * Key insight: During a real blink, the iris is hidden and doesn't move.
 * During a gaze shift, the iris moves significantly while EAR changes.
 */

import { IRIS_INDICES, EYE_LID_INDICES, GAZE_TRACKING } from './constants';
import { Point2D } from './blink';

export interface GazeResult {
  /** Iris Y position ratio (0=top of eye, 1=bottom, 0.5=centered) */
  leftIrisRatio: number;
  /** Iris Y position ratio for right eye */
  rightIrisRatio: number;
  /** Average iris ratio (both eyes) */
  avgIrisRatio: number;
  /** Smoothed iris ratio (EMA filtered) */
  smoothedIrisRatio: number;
  /** Current iris velocity (ratio units per frame, positive = looking down) */
  velocity: number;
  /** True if significant gaze shift detected in recent window */
  isGazeShift: boolean;
  /** Total displacement in detection window */
  recentDisplacement: number;
  /** True if tracking is reliable (eyes sufficiently open) */
  isReliable: boolean;
}

interface GazeSample {
  irisRatio: number;
  timestamp: number;
}

/**
 * Tracks iris position to detect vertical gaze shifts (looking up/down).
 *
 * Usage:
 *   const tracker = new GazeTracker();
 *
 *   for (const frame of cameraFrames) {
 *     const landmarks = mediapipe.detect(frame);
 *     const result = tracker.update(landmarks, timestamp);
 *
 *     if (result.isGazeShift) {
 *       // Suppress blink detection - this is a gaze movement, not a blink
 *     }
 *   }
 */
export class GazeTracker {
  // Smoothed iris position
  private smoothedIrisRatio: number = 0.5;

  // History for velocity calculation
  private history: GazeSample[] = [];

  // Configuration
  private readonly velocityWindowFrames: number;
  private readonly velocityThreshold: number;
  private readonly detectionWindowMs: number;
  private readonly displacementThreshold: number;
  private readonly smoothingAlpha: number;
  private readonly minEyeOpeningPixels: number;
  private readonly enabled: boolean;

  constructor(config?: Partial<typeof GAZE_TRACKING>) {
    const cfg = { ...GAZE_TRACKING, ...config };
    this.velocityWindowFrames = cfg.VELOCITY_WINDOW_FRAMES;
    this.velocityThreshold = cfg.VELOCITY_THRESHOLD;
    this.detectionWindowMs = cfg.DETECTION_WINDOW_MS;
    this.displacementThreshold = cfg.MIN_DISPLACEMENT_THRESHOLD;
    this.smoothingAlpha = cfg.SMOOTHING_ALPHA;
    this.minEyeOpeningPixels = cfg.MIN_EYE_OPENING_PIXELS;
    this.enabled = cfg.ENABLED;
  }

  /**
   * Calculate iris Y position ratio within the eye opening.
   *
   * @param iris Iris center landmark
   * @param upperLid Upper eyelid center landmark
   * @param lowerLid Lower eyelid center landmark
   * @returns Ratio from 0 (top) to 1 (bottom), 0.5 is centered
   */
  private calculateIrisRatio(
    iris: Point2D,
    upperLid: Point2D,
    lowerLid: Point2D
  ): { ratio: number; eyeOpening: number } {
    const eyeOpening = lowerLid.y - upperLid.y;

    // If eye is closed or nearly closed, ratio is unreliable
    if (eyeOpening <= 0) {
      return { ratio: 0.5, eyeOpening: 0 };
    }

    // Calculate where iris is within the eye opening
    const irisPosition = iris.y - upperLid.y;
    const ratio = irisPosition / eyeOpening;

    // Clamp to valid range (iris shouldn't be outside eye)
    const clampedRatio = Math.max(0, Math.min(1, ratio));

    return { ratio: clampedRatio, eyeOpening };
  }

  /**
   * Update tracker with new frame landmarks.
   *
   * @param landmarks Array of 478 MediaPipe face landmarks
   * @param timestampMs Current timestamp in milliseconds
   * @returns GazeResult with current gaze state
   */
  update(landmarks: Point2D[], timestampMs?: number): GazeResult {
    const ts = timestampMs ?? performance.now();

    // Default result for disabled or unreliable tracking
    const defaultResult: GazeResult = {
      leftIrisRatio: 0.5,
      rightIrisRatio: 0.5,
      avgIrisRatio: 0.5,
      smoothedIrisRatio: this.smoothedIrisRatio,
      velocity: 0,
      isGazeShift: false,
      recentDisplacement: 0,
      isReliable: false,
    };

    if (!this.enabled) {
      return defaultResult;
    }

    // Extract landmarks
    const leftIris = landmarks[IRIS_INDICES.LEFT_CENTER];
    const rightIris = landmarks[IRIS_INDICES.RIGHT_CENTER];
    const leftUpperLid = landmarks[EYE_LID_INDICES.LEFT_UPPER];
    const leftLowerLid = landmarks[EYE_LID_INDICES.LEFT_LOWER];
    const rightUpperLid = landmarks[EYE_LID_INDICES.RIGHT_UPPER];
    const rightLowerLid = landmarks[EYE_LID_INDICES.RIGHT_LOWER];

    // Validate landmarks exist
    if (!leftIris || !rightIris || !leftUpperLid || !leftLowerLid ||
        !rightUpperLid || !rightLowerLid) {
      return defaultResult;
    }

    // Calculate iris ratios for each eye
    const leftResult = this.calculateIrisRatio(leftIris, leftUpperLid, leftLowerLid);
    const rightResult = this.calculateIrisRatio(rightIris, rightUpperLid, rightLowerLid);

    // Check if eyes are open enough for reliable tracking
    const minOpening = Math.min(leftResult.eyeOpening, rightResult.eyeOpening);
    const isReliable = minOpening >= this.minEyeOpeningPixels;

    if (!isReliable) {
      // Eyes too closed - don't update smoothed values, just return current state
      return {
        ...defaultResult,
        leftIrisRatio: leftResult.ratio,
        rightIrisRatio: rightResult.ratio,
        avgIrisRatio: (leftResult.ratio + rightResult.ratio) / 2,
        isReliable: false,
      };
    }

    // Average both eyes for robustness
    const avgIrisRatio = (leftResult.ratio + rightResult.ratio) / 2;

    // Apply EMA smoothing
    this.smoothedIrisRatio =
      this.smoothingAlpha * avgIrisRatio +
      (1 - this.smoothingAlpha) * this.smoothedIrisRatio;

    // Add to history
    this.history.push({ irisRatio: this.smoothedIrisRatio, timestamp: ts });

    // Prune old history (keep enough for velocity + displacement calculation)
    const cutoffTime = ts - this.detectionWindowMs;
    while (this.history.length > 0 && this.history[0].timestamp < cutoffTime) {
      this.history.shift();
    }

    // Calculate velocity (change over last N frames)
    let velocity = 0;
    if (this.history.length >= 2) {
      const recent = this.history.slice(-this.velocityWindowFrames);
      if (recent.length >= 2) {
        const oldest = recent[0];
        const newest = recent[recent.length - 1];
        const frames = recent.length - 1;
        velocity = (newest.irisRatio - oldest.irisRatio) / frames;
      }
    }

    // Calculate total displacement in detection window
    let recentDisplacement = 0;
    if (this.history.length >= 2) {
      // Find min and max in window
      let minRatio = Infinity;
      let maxRatio = -Infinity;
      for (const sample of this.history) {
        minRatio = Math.min(minRatio, sample.irisRatio);
        maxRatio = Math.max(maxRatio, sample.irisRatio);
      }
      recentDisplacement = maxRatio - minRatio;
    }

    // Detect gaze shift: high velocity OR large displacement
    const isGazeShift =
      Math.abs(velocity) > this.velocityThreshold ||
      recentDisplacement > this.displacementThreshold;

    return {
      leftIrisRatio: leftResult.ratio,
      rightIrisRatio: rightResult.ratio,
      avgIrisRatio,
      smoothedIrisRatio: this.smoothedIrisRatio,
      velocity,
      isGazeShift,
      recentDisplacement,
      isReliable,
    };
  }

  /**
   * Check if a gaze shift occurred recently (for blink rejection).
   * This is a convenience method for use in blink detection.
   *
   * @returns True if significant gaze movement detected in recent window
   */
  hasRecentGazeShift(): boolean {
    if (this.history.length < 2) {
      return false;
    }

    // Check displacement in window
    let minRatio = Infinity;
    let maxRatio = -Infinity;
    for (const sample of this.history) {
      minRatio = Math.min(minRatio, sample.irisRatio);
      maxRatio = Math.max(maxRatio, sample.irisRatio);
    }
    const displacement = maxRatio - minRatio;

    return displacement > this.displacementThreshold;
  }

  /**
   * Get current gaze direction as a descriptive string.
   */
  getGazeDirection(): 'up' | 'center' | 'down' {
    if (this.smoothedIrisRatio < 0.4) {
      return 'up';
    } else if (this.smoothedIrisRatio > 0.6) {
      return 'down';
    }
    return 'center';
  }

  /**
   * Reset tracker state.
   */
  reset(): void {
    this.smoothedIrisRatio = 0.5;
    this.history = [];
  }

  /**
   * Check if gaze tracking is enabled.
   */
  isEnabled(): boolean {
    return this.enabled;
  }
}

File: experiments/robust_blink_v2/gaze_tracker.py
"""
Gaze tracking for detecting vertical eye movements (looking up/down).

Tracks iris position relative to eye opening to detect gaze shifts.
Used to gate blink detection - rapid look down/up causes EAR changes that
can be mistaken for blinks.

The key insight: When you look down, your iris moves toward the lower eyelid.
When you blink, your eyelids close over your iris - the iris doesn't move.
By tracking iris position, we can distinguish the two.
"""

from dataclasses import dataclass
from typing import Optional, List, Tuple
from collections import deque
import time

from constants import (
    # Iris landmarks
    LEFT_IRIS_CENTER_INDEX,
    RIGHT_IRIS_CENTER_INDEX,
    # Eyelid landmarks
    LEFT_UPPER_LID_INDEX,
    LEFT_LOWER_LID_INDEX,
    RIGHT_UPPER_LID_INDEX,
    RIGHT_LOWER_LID_INDEX,
    # Gaze tracking parameters
    GAZE_TRACKING_ENABLED,
    GAZE_VELOCITY_WINDOW_FRAMES,
    GAZE_VELOCITY_THRESHOLD,
    GAZE_DETECTION_WINDOW_MS,
    GAZE_MIN_DISPLACEMENT_THRESHOLD,
    GAZE_SMOOTHING_ALPHA,
    GAZE_MIN_EYE_OPENING,
)


@dataclass
class GazeResult:
    """Result of gaze tracking analysis."""
    # Raw iris ratios (0 = upper lid, 1 = lower lid, 0.5 = centered)
    left_iris_ratio: float
    right_iris_ratio: float
    avg_iris_ratio: float

    # Smoothed ratio (EMA applied)
    smoothed_iris_ratio: float

    # Velocity (change per frame, positive = looking down)
    velocity: float

    # Total displacement in detection window
    recent_displacement: float

    # Detection flags
    is_gaze_shift: bool      # True if rapid vertical movement detected
    is_reliable: bool        # False if eyes too closed for reliable tracking

    # Gaze direction for debugging
    gaze_direction: str      # "up", "center", or "down"


class GazeTracker:
    """
    Tracks vertical iris position to detect looking up/down movements.

    Algorithm:
    1. Calculate iris Y position relative to eye opening (0=top, 1=bottom)
    2. Apply EMA smoothing to reduce noise
    3. Track velocity (change between frames)
    4. Track total displacement over detection window
    5. Flag as gaze shift if velocity OR displacement exceeds threshold
    """

    def __init__(
        self,
        enabled: bool = GAZE_TRACKING_ENABLED,
        velocity_threshold: float = GAZE_VELOCITY_THRESHOLD,
        displacement_threshold: float = GAZE_MIN_DISPLACEMENT_THRESHOLD,
        smoothing_alpha: float = GAZE_SMOOTHING_ALPHA,
        min_eye_opening: float = GAZE_MIN_EYE_OPENING,
        velocity_window: int = GAZE_VELOCITY_WINDOW_FRAMES,
        detection_window_ms: float = GAZE_DETECTION_WINDOW_MS,
    ):
        self.enabled = enabled
        self.velocity_threshold = velocity_threshold
        self.displacement_threshold = displacement_threshold
        self.smoothing_alpha = smoothing_alpha
        self.min_eye_opening = min_eye_opening
        self.velocity_window = velocity_window
        self.detection_window_ms = detection_window_ms

        # State
        self.smoothed_ratio: float = 0.5  # Start centered
        self.prev_ratio: float = 0.5

        # History for velocity/displacement calculation
        # Each entry: (timestamp_ms, iris_ratio)
        self.history: deque = deque(maxlen=100)  # More than enough

        # Recent gaze shift flag (for hasRecentGazeShift)
        self.recent_gaze_shift: bool = False

    def update(
        self,
        landmarks: list,
        timestamp_ms: Optional[float] = None,
    ) -> GazeResult:
        """
        Update gaze tracking with new landmarks.

        Args:
            landmarks: List of MediaPipe NormalizedLandmark objects
            timestamp_ms: Frame timestamp in milliseconds

        Returns:
            GazeResult with iris tracking data
        """
        if timestamp_ms is None:
            timestamp_ms = time.time() * 1000

        # Return default result if disabled
        if not self.enabled:
            return GazeResult(
                left_iris_ratio=0.5,
                right_iris_ratio=0.5,
                avg_iris_ratio=0.5,
                smoothed_iris_ratio=0.5,
                velocity=0.0,
                recent_displacement=0.0,
                is_gaze_shift=False,
                is_reliable=False,
                gaze_direction="center",
            )

        # Calculate iris ratio for each eye
        left_ratio, left_reliable = self._calculate_iris_ratio(
            landmarks,
            LEFT_IRIS_CENTER_INDEX,
            LEFT_UPPER_LID_INDEX,
            LEFT_LOWER_LID_INDEX,
        )

        right_ratio, right_reliable = self._calculate_iris_ratio(
            landmarks,
            RIGHT_IRIS_CENTER_INDEX,
            RIGHT_UPPER_LID_INDEX,
            RIGHT_LOWER_LID_INDEX,
        )

        # Average both eyes
        avg_ratio = (left_ratio + right_ratio) / 2.0
        is_reliable = left_reliable and right_reliable

        # Apply EMA smoothing
        self.smoothed_ratio = (
            self.smoothing_alpha * avg_ratio +
            (1 - self.smoothing_alpha) * self.smoothed_ratio
        )

        # Calculate velocity (change from previous frame)
        velocity = self.smoothed_ratio - self.prev_ratio
        self.prev_ratio = self.smoothed_ratio

        # Add to history
        self.history.append((timestamp_ms, self.smoothed_ratio))

        # Calculate displacement over detection window
        recent_displacement = self._calculate_recent_displacement(timestamp_ms)

        # Detect gaze shift
        # Either high velocity (fast movement) OR large displacement (gradual movement)
        is_gaze_shift = False
        if is_reliable:
            high_velocity = abs(velocity) > self.velocity_threshold
            large_displacement = recent_displacement > self.displacement_threshold
            is_gaze_shift = high_velocity or large_displacement

        self.recent_gaze_shift = is_gaze_shift

        # Determine gaze direction for debugging
        gaze_direction = "center"
        if self.smoothed_ratio > 0.6:
            gaze_direction = "down"
        elif self.smoothed_ratio < 0.4:
            gaze_direction = "up"

        return GazeResult(
            left_iris_ratio=left_ratio,
            right_iris_ratio=right_ratio,
            avg_iris_ratio=avg_ratio,
            smoothed_iris_ratio=self.smoothed_ratio,
            velocity=velocity,
            recent_displacement=recent_displacement,
            is_gaze_shift=is_gaze_shift,
            is_reliable=is_reliable,
            gaze_direction=gaze_direction,
        )

    def _calculate_iris_ratio(
        self,
        landmarks: list,
        iris_index: int,
        upper_lid_index: int,
        lower_lid_index: int,
    ) -> Tuple[float, bool]:
        """
        Calculate iris Y position as ratio within eye opening.

        Returns:
            (ratio, is_reliable): ratio is 0-1 (0=top, 1=bottom),
                                  is_reliable is False if eye too closed
        """
        iris = landmarks[iris_index]
        upper_lid = landmarks[upper_lid_index]
        lower_lid = landmarks[lower_lid_index]

        # Eye opening (vertical distance between lids)
        eye_opening = lower_lid.y - upper_lid.y

        # Check if eye is open enough for reliable tracking
        if eye_opening < self.min_eye_opening:
            return 0.5, False  # Return centered when unreliable

        # Calculate ratio: where is iris between upper (0) and lower (1) lid?
        iris_ratio = (iris.y - upper_lid.y) / eye_opening

        # Clamp to 0-1 range (iris should always be between lids)
        iris_ratio = max(0.0, min(1.0, iris_ratio))

        return iris_ratio, True

    def _calculate_recent_displacement(self, current_ms: float) -> float:
        """
        Calculate total iris movement over detection window.

        Looks at max - min ratio within the window to catch
        oscillating movements (look down then up).
        """
        if len(self.history) < 2:
            return 0.0

        # Find entries within detection window
        window_start = current_ms - self.detection_window_ms
        recent_ratios = [
            ratio for ts, ratio in self.history
            if ts >= window_start
        ]

        if len(recent_ratios) < 2:
            return 0.0

        # Displacement = max - min (catches full movement range)
        return max(recent_ratios) - min(recent_ratios)

    def has_recent_gaze_shift(self) -> bool:
        """Check if a gaze shift was detected in the most recent update."""
        return self.recent_gaze_shift

    def get_gaze_direction(self) -> str:
        """Get current gaze direction as string."""
        if self.smoothed_ratio > 0.6:
            return "down"
        elif self.smoothed_ratio < 0.4:
            return "up"
        return "center"

    def is_enabled(self) -> bool:
        """Check if gaze tracking is enabled."""
        return self.enabled

    def reset(self):
        """Reset tracking state."""
        self.smoothed_ratio = 0.5
        self.prev_ratio = 0.5
        self.history.clear()
        self.recent_gaze_shift = False
